From 4a292a41b74fcd9ef283a4eff0f191a3b94e66c8 Mon Sep 17 00:00:00 2001
From: Alex Marginean <alexandru.marginean@nxp.com>
Date: Wed, 7 Feb 2018 17:30:19 +0200
Subject: [PATCH 106/706] enetc: set up SI cache attributes

These are the attributes used by ENETC hardware when accessing memory and
issuing MSI-Xs, specifically coherency and cache allocation.

Signed-off-by: Alex Marginean <alexandru.marginean@nxp.com>
(cherry picked from commit eda214e0ab8ed2f0f6b2c5eba1051856abe8c271)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/ethernet/freescale/enetc/enetc.c    | 5 ++++-
 drivers/net/ethernet/freescale/enetc/enetc_hw.h | 7 +++++++
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/freescale/enetc/enetc.c b/drivers/net/ethernet/freescale/enetc/enetc.c
index 816003f6391c..20bfe6f77516 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc.c
@@ -892,7 +892,10 @@ static void enetc_configure_si(struct enetc_si *si)
 	struct enetc_hw *hw = &si->hw;
 
 	enetc_setup_cbdr(hw, &si->cbd_ring);
-
+	/* set SI cache attributes */
+	enetc_wr(hw, ENETC_SICAR0,
+		 ENETC_SICAR_RD_COHERENT | ENETC_SICAR_WR_COHERENT);
+	enetc_wr(hw, ENETC_SICAR1, ENETC_SICAR_WR_MSI);
 	/* enable SI, start RSS by default */
 	enetc_wr(hw, ENETC_SIMR, ENETC_SIMR_EN | ENETC_SIMR_RSSE);
 }
diff --git a/drivers/net/ethernet/freescale/enetc/enetc_hw.h b/drivers/net/ethernet/freescale/enetc/enetc_hw.h
index b88a2bc22092..b8683dbb0f97 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc_hw.h
+++ b/drivers/net/ethernet/freescale/enetc/enetc_hw.h
@@ -46,6 +46,13 @@
 #define ENETC_SIMR_EN	BIT(31)
 #define ENETC_SIMR_RSSE	BIT(0)
 
+/* Cache attribute registers for transactions initiated by ENETC */
+#define ENETC_SICAR0	0x08
+#define ENETC_SICAR1	0x0C
+#define ENETC_SICAR_RD_COHERENT	0x2b2b0000 /* rd snoop, no alloc */
+#define ENETC_SICAR_WR_COHERENT	0x00006767 /* wr snoop, no alloc */
+#define ENETC_SICAR_WR_MSI	0x00000037 /* wr no snoop, no alloc */
+
 #define ENETC_SIPMAR0	0x80
 #define ENETC_SIPMAR1	0x84
 
-- 
2.17.1

